### ISSUE TITLE
multi-layer issues

### ISSUE THREAD

**Issue by sdivitt** [bahremsd/tmmax#44]
Thank you for your work on this project. I have run into an issue that you might consider looking into.

Here is an example where I run tmm with two apparently different stacks but where the results should be identical. If I try ["Air", "SiO2", "Air"] and ["Air", "Air", "Air", "Air", "SiO2", "Air"] I find different results. I believe the results should actually be the same because there will be no reflections between the "Air" layers, they should act as a single layer. Example code:

<img width="1000" height="600" alt="Image" src="https://github.com/user-attachments/assets/7baf16be-e237-4225-a292-331ef695b0f2" />

```
import jax.numpy as jnp
import matplotlib.pyplot as plt
from tmmax.tmm import tmm

# ==========================================
# 1. TMMAX SIMULATION (User Code)
# ==========================================

# Define parameters
material_list = ["Air", "Air", "Air", "Air", "SiO2", "Air"]
thickness_list = jnp.array([100e-9,100e-9,100e-9,500e-9])
wavelength_arr = jnp.linspace(400e-9, 800e-9, 1000)
angle_of_incidences = jnp.array([45 * jnp.pi / 180])

# Run S-Polarization
polarization = 's'
R_s_sim, T_s_sim = tmm(material_list=material_list,
                       thickness_list=thickness_list,
                       wavelength_arr=wavelength_arr,
                       angle_of_incidences=angle_of_incidences,
                       polarization=polarization)

# Run P-Polarization
polarization = 'p'
R_p_sim, T_p_sim = tmm(material_list=material_list,
                       thickness_list=thickness_list,
                       wavelength_arr=wavelength_arr,
                       angle_of_incidences=angle_of_incidences,
                       polarization=polarization)

# ==========================================
# 2. TMMAX SIMULATION (User Code)
# ==========================================

# Define parameters
material_list = ["Air", "SiO2", "Air"]
thickness_list = jnp.array([500e-9])
wavelength_arr = jnp.linspace(400e-9, 800e-9, 1000)
angle_of_incidences = jnp.array([45 * jnp.pi / 180])

# Run S-Polarization
polarization = 's'
R_s_sim2, T_s_sim2 = tmm(material_list=material_list,
                       thickness_list=thickness_list,
                       wavelength_arr=wavelength_arr,
                       angle_of_incidences=angle_of_incidences,
                       polarization=polarization)

# Run P-Polarization
polarization = 'p'
R_p_sim2, T_p_sim2 = tmm(material_list=material_list,
                       thickness_list=thickness_list,
                       wavelength_arr=wavelength_arr,
                       angle_of_incidences=angle_of_incidences,
                       polarization=polarization)


# ==========================================
# 3. PLOTTING
# ==========================================

plt.figure(figsize=(10, 6))

# Plot S-Polarization
plt.plot(wavelength_arr * 1e9, R_s_sim[0, :], 'o', markersize=4, label='6 layer (s-pol)', color='blue', alpha=0.5)
plt.plot(wavelength_arr * 1e9, R_s_sim2[0, :], 'x', markersize=4, label='3 layer (s-pol)', color='purple', alpha=0.5)

# Plot P-Polarization
plt.plot(wavelength_arr * 1e9, R_p_sim[0, :], 'o', markersize=4, label='6 layer (p-pol)', color='red', alpha=0.5)
plt.plot(wavelength_arr * 1e9, R_p_sim2[0, :], 'o', markersize=4, label='3 layer (p-pol)', color='green', alpha=0.5)

plt.xlabel("Wavelength (nm)")
plt.ylabel("Reflectivity")
plt.legend()
plt.grid(True, alpha=0.3)
plt.show()
```

**Comment by bahremsd** [comment]
Hello,

Thank you very much for pointing this out. You are absolutely right about that, in a linear system, consecutive identical layers (e.g., multiple adjacent “Air” layers) should behave as a single effective layer.

When designing the JIT-compatible version (fixed shape array management etc.) of tmm, we made an implicit assumption (sorry for not mentioning in the documentation): practical users and designers typically avoid writing consecutive identical materials and instead merge them manually into a single layer. Because of that assumption, the current implementation does not automatically merge adjacent identical materials before running the JIT-compiled TMM routine. This is the reason you observe a difference between:

- ["Air", "SiO2", "Air"]

- ["Air", "Air", "Air", "Air", "SiO2", "Air"]

The correct behavior can be restored by merging the material list and thickness list before calling the JIT-compiled TMM function. We will incorporate an automatic merging procedure in the upcoming release so that the user does not need to handle this manually. Until then, merging identical neighboring layers before running the simulation produces the expected physically consistent results.

For reference, we have already demonstrated the accuracy of the merged-layer version of tmmax compared with the standard tmm package in the following issue:
https://github.com/bahremsd/tmmax/issues/19

I am closing this issue for now. The fix will be included in next month’s release, and this behavior will no longer occur.

Thank you again for your helpful feedback.

**Comment by bahremsd** [comment]
You are absolutely right that, from a purely mathematical standpoint, consecutive identical layers should produce the same reflectivity whether they are merged or not.

In our implementation, however, we introduced this “no adjacent identical layers” assumption intentionally as a performance optimization. The goal was to keep the number of internal computations as low as possible to maximize speed and JIT efficiency. Because of this design choice, the function currently expects users not to include back-to-back identical materials, and it does not automatically merge them.

That said, we fully acknowledge your point: ideally, the results should be identical with or without merging, and the discrepancy you observed indicates a deeper issue that we will investigate. We will work on a solution in a future release that preserves both physical correctness and computational performance.

Thank you again for your insightful feedback. Your input is very helpful for improving the package.
